### PR TITLE
Integrate NVIDIA's S/R implementation into HPX 

### PR DESCRIPTION
**Pull request by isidorostsa**
This PR serves to replace the existing senders/receivers implementation in hpx with the [stdexec implementation](https://github.com/NVIDIA/stdexec) maintained by NVIDIA.

Big credits to the @pika-org team for their work when solving almost the same problem. This PR is built on top of much of that work.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -2196,6 +2196,9 @@ include(HPX_SetupAsio)
 include(HPX_SetupAllocator)
 include(HPX_SetupHwloc)
 
+# Setup NVIDIA's stdexec
+include(HPX_SetupStdexec)
```
Comment (hkaiser) [inline CMakeLists.txt:2200]: This probably should be included only if compiling with C++20 or higher. We should also introduce an `HPX_WITH_...` option to explicitly enable pulling it in.

Comment (isidorostsa) [reply CMakeLists.txt:2200]: Yess totally, and more nuances are to come 
(this is a bare minimum pr for now)

File: CMakeLists.txt [hunk 2]
```diff
@@ -567,6 +567,49 @@ if(HPX_WITH_CUDA AND HPX_WITH_HIP)
...
+  HPX_WITH_FETCH_STDEXEC
+  BOOL
+  "Use FetchContent to fetch STDEXEC.(default: OFF)"
+  OFF
+  CATEGORY "Executor"
+  ADVANCED
+)
+
+# if STDEXEC_ROOT was provided, HPX_WITH_STDEXEC is set to ON
+if(STDEXEC_ROOT OR HPX_WITH_FETCH_STDEXEC)
```
Comment (hkaiser) [inline CMakeLists.txt:593]: This might have to be `Stdexec_ROOT` to avoid warnings with recent CMake versions

File: cmake/FindStdexec.cmake [hunk 3]
```diff
@@ -0,0 +1,33 @@
+
```
Comment (hkaiser) [inline cmake/FindStdexec.cmake:1]: Please add Copyright, License and the SPDX identifier to newly created files.

File: cmake/HPX_SetupStdexec.cmake [hunk 4]
```diff
@@ -0,0 +1,34 @@
+if(STDEXEC_ROOT AND NOT Stdexec_ROOT)
+    set(Stdexec_ROOT
+        ${STDEXEC_ROOT}
+        CACHE PATH "STDEXEC base directory"
+    )
+    unset(STDEXEC_ROOT CACHE)
+endif()
```
Comment (hkaiser) [inline cmake/HPX_SetupStdexec.cmake:7]: We have this kind of code in our CMake build system only to be backwards compatible, not sure if we need this at all.

File: cmake/HPX_SetupStdexec.cmake [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+if(HPX_WITH_FETCH_STDEXEC AND NOT Stdexec_ROOT)
+    hpx_info(
+        "HPX_WITH_FETCH_STDEXEC=${HPX_WITH_FETCH_STDEXEC}, Stdexec will be fetched using CMake's FetchContent."
+    )
+    if(UNIX)
+        include(FetchContent)
+        FetchContent_Declare(
+            Stdexec
+            GIT_REPOSITORY https://github.com/NVIDIA/stdexec.git
+            GIT_TAG        main
```
Comment (hkaiser) [inline cmake/HPX_SetupStdexec.cmake:18]: Can we make the fetched tag configurable, please?

File: libs/CMakeLists.txt [hunk 6]
```diff
@@ -113,6 +113,11 @@ if(TARGET APEX::apex)
   target_link_libraries(hpx_full INTERFACE APEX::apex)
 endif()
 
+# Do I need this here?
```
Comment (hkaiser) [inline libs/CMakeLists.txt:116]: You most likely will not need this as it will be propagated from `hpx_core`.

File: libs/core/execution_base/CMakeLists.txt [hunk 7]
```diff
@@ -52,6 +55,8 @@ add_hpx_module(
   SOURCES ${execution_base_sources}
   HEADERS ${execution_base_headers}
   COMPAT_HEADERS ${execution_base_compat_headers}
+  DEPENDENCIES
+    STDEXEC::stdexec
```
Comment (hkaiser) [inline libs/core/execution_base/CMakeLists.txt:59]: This is needed only if enabled.

File: cmake/HPX_SetupStdexec.cmake (hunk 5 above)
Comment (isidorostsa) [reply cmake/HPX_SetupStdexec.cmake:18]: Of course!

File: cmake/HPX_SetupStdexec.cmake (hunk 4 above)
Comment (isidorostsa) [reply cmake/HPX_SetupStdexec.cmake:7]: Makes sense, removing

File: CMakeLists.txt (hunk 1 above)
Comment (isidorostsa) [reply CMakeLists.txt]: Do you think it would make sense to perform this check when parsing the arguments in the [main CMakeLists.txt](https://github.com/STEllAR-GROUP/hpx/pull/6431/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR611)?

I did not spot any convention for that sort of thing in the CMake files.

Comment (hkaiser) [reply CMakeLists.txt]: > Do you think it would make sense to perform this check when parsing the arguments in the [main CMakeLists.txt](https://github.com/STEllAR-GROUP/hpx/pull/6431/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR611)?
> 
> I did not spot any convention for that sort of thing in the CMake files.

Yah, we don't have conventions. Let's move as much as possible out of the main CMakeLists.txt, it's overcrowded already.

File: libs/CMakeLists.txt (hunk 6 above)
Comment (isidorostsa) [reply libs/CMakeLists.txt:116]: Removed

File: CMakeLists.txt (hunk 2 above)
Comment (isidorostsa) [reply CMakeLists.txt]: Changed it, thank you

File: libs/core/execution/include/hpx/execution/algorithms/as_sender.hpp [hunk 8]
```diff
@@ -136,6 +136,27 @@ namespace hpx::execution::experimental {
 
             std::decay_t<Future> future_;
 
+#ifdef HPX_HAVE_STDEXEC
+            template <bool IsVoid, typename _result_type>
+            struct _set_value
```
Comment (hkaiser) [inline libs/core/execution/include/hpx/execution/algorithms/as_sender.hpp:141]: Let's not start with the leading underscore madness...

File: libs/core/execution/include/hpx/execution/queries/read.hpp [hunk 9]
```diff
@@ -87,7 +92,7 @@ namespace hpx::execution::experimental {
             template <typename Tag1>
             friend auto tag_invoke(
                 get_completion_signatures_t, read_sender<Tag1>, no_env)
-                -> dependent_completion_signatures<no_env>;
+                -> std::execution::dependent_completion_signatures<no_env>;
```
Comment (hkaiser) [inline libs/core/execution/include/hpx/execution/queries/read.hpp:95]: this doesn't look right, is std::execution always available?

File: libs/core/execution/tests/unit/algorithm_as_sender.cpp [hunk 10]
```diff
@@ -30,7 +30,11 @@ int hpx_main()
         auto s = ex::as_sender(std::move(fut));
 
         static_assert(ex::is_sender_v<decltype(s)>);
+#ifdef HPX_HAVE_STDEXEC
+        static_assert(ex::is_sender_in_v<decltype(s), ex::empty_env>);
+#else
         static_assert(ex::is_sender_v<decltype(s), ex::empty_env>);
+#endif
```
Comment (hkaiser) [inline libs/core/execution/tests/unit/algorithm_as_sender.cpp:37]: Shouldn't we rather implement `is_sender_in_v` for our code base?

**Review by hkaiser — COMMENTED** [review]
First batch of comments... more to come.

File: libs/core/execution/include/hpx/execution/algorithms/as_sender.hpp (hunk 8 above)
Comment (isidorostsa) [reply libs/core/execution/include/hpx/execution/algorithms/as_sender.hpp:141]: right, will rename to `set_value_void_checked` or something similar. I've implemented something this in other parts of the code so I will make sure the names match. I do not think it is worth it to factor this out since it's only 2 or 3 places.

File: libs/core/execution/include/hpx/execution/algorithms/detail/partial_algorithm.hpp [hunk 11]
```diff
@@ -62,9 +62,16 @@ namespace hpx::execution::experimental::detail {
...
+#ifdef HPX_HAVE_STDEXEC
+            auto scheduler =
+                hpx::execution::experimental::get_completion_scheduler<
+                    hpx::execution::experimental::set_value_t>(
+                    hpx::execution::experimental::get_env(u));
+#else
             auto scheduler =
                 hpx::execution::experimental::get_completion_scheduler<
                     hpx::execution::experimental::set_value_t>(u);
+#endif
```
[the remark this answers is not in the document]
Comment (isidorostsa) [reply libs/core/execution/include/hpx/execution/algorithms/detail/partial_algorithm.hpp:74]: We would have to alter our existing senders to provide environments, currently most of the times they do not.

File: libs/core/algorithms/tests/unit/algorithms/adjacentdifference_tests.hpp [hunk 12]
```diff
@@ -83,8 +84,14 @@ void test_adjacent_difference_sender(Policy l, ExPolicy&& policy)
...
 
     auto exec = ex::explicit_scheduler_executor(scheduler_t(l));
+#ifdef HPX_HAVE_STDEXEC
+    auto result =
+        tt::sync_wait(ex::just(std::begin(c), std::end(c), std::begin(d)) |
+            hpx::adjacent_difference(policy.on(exec)));
+#else
     auto result = ex::just(std::begin(c), std::end(c), std::begin(d)) |
         hpx::adjacent_difference(policy.on(exec)) | tt::sync_wait();
+#endif
```
[the remark this answers is not in the document]
Comment (isidorostsa) [reply libs/core/algorithms/tests/unit/algorithms/adjacentdifference_tests.hpp:94]: that duplication is present in many of the tests i have changed so it might make sense to do that there too?
e.g. https://github.com/STEllAR-GROUP/hpx/pull/6431/files#diff-b11a40f80e11363af9d5469224fdb4e94487ee5ff99c30c7d4219a5ab40c85d8

File: libs/core/execution/tests/unit/algorithm_as_sender.cpp (hunk 10 above)
Comment (isidorostsa) [reply libs/core/execution/tests/unit/algorithm_as_sender.cpp:37]: right, leaving that as a TODO to remove much of the duplication from the tests.

File: libs/core/execution/include/hpx/execution/algorithms/let_stopped.hpp [hunk 13]
```diff
@@ -8,6 +8,9 @@
 #pragma once
 
 #include <hpx/config.hpp>
+#ifdef HPX_HAVE_STDEXEC
+#else
```
[the remark this answers is not in the document]
Comment (isidorostsa) [reply libs/core/execution/include/hpx/execution/algorithms/let_stopped.hpp:12]: ah yes, thanks!

File: libs/core/execution/include/hpx/execution/queries/read.hpp (hunk 9 above)
Comment (isidorostsa) [reply libs/core/execution/include/hpx/execution/queries/read.hpp:95]: You are right, it is not and this is wrong. Also, this part of the code is ifdef-ed away so I am reverting it to the pre-stdexec code.

File: libs/core/execution/include/hpx/execution/algorithms/detail/partial_algorithm.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (hkaiser) [reply libs/core/execution/include/hpx/execution/algorithms/detail/partial_algorithm.hpp:74]: Really? What about:
```
inline constexpr struct get_completion_scheduler_ex_t
{
    template <typename U>
    decltype(auto) operator()(U&& u) const
    {
#ifdef HPX_HAVE_STDEXEC
        return hpx::execution::experimental::get_completion_scheduler<
                    hpx::execution::experimental::set_value_t>(
                    hpx::execution::experimental::get_env(HPX_FORWARD(U, u)));
#else
        return hpx::execution::experimental::get_completion_scheduler<
                    hpx::execution::experimental::set_value_t>(HPX_FORWARD(U, u));
#endif
    }
} get_completion_scheduler_ex{};
```

**Comment by isidorostsa** [comment]
Fetching STDEXEC by default on C++20, so the CI is testing both with and without it.

File: libs/core/execution/tests/unit/algorithm_run_loop.cpp [hunk 14]
```diff
@@ -761,23 +760,6 @@ void test_future_sender()
...
-        HPX_TEST_EQ(f.get(), 3);
-    }
-    // Make a future from a scheduler
-    {
-        ex::run_loop loop;
-        auto sched = loop.get_scheduler();
-
-        auto f = ex::make_future(sched);
-    }
-
```
Comment (hkaiser) [inline libs/core/execution/tests/unit/algorithm_run_loop.cpp]: Why did you remove these tests?

Comment (isidorostsa) [reply libs/core/execution/tests/unit/algorithm_run_loop.cpp]: I added those tests at some point but I think because I was trying to understand how `make_future` works.
I removed them because the first one is pretty much a duplicate of  https://github.com/STEllAR-GROUP/hpx/pull/6431/commits/ef31254b19a6b730ca8863df58b9b95c3e79e1ec#diff-b11a40f80e11363af9d5469224fdb4e94487ee5ff99c30c7d4219a5ab40c85d8L799 and the second one is using `make_future` on a scheduler instead of a sender which is not correct in the first place.

**Comment by hkaiser** [comment]
@isidorostsa just in case you were wondering... the Windows tests all fail with a segmentation fault (error code `0xc0000005`)

**Comment by isidorostsa** [comment]
> @isidorostsa just in case you were wondering... the Windows tests all fail with a segmentation fault (error code `0xc0000005`)

Interestingly it's using c++17 too

File: libs/core/async_mpi/tests/unit/algorithm_transform_mpi.cpp [hunk 15]
```diff
@@ -212,8 +212,11 @@ int hpx_main()
                 }
                 catch (std::runtime_error const& e)
                 {
-                    HPX_TEST(std::string(e.what()).find(std::string(
-                                 "invalid root")) != std::string::npos);
+                    auto error_msg = std::string(e.what());
+                    for (auto& c : error_msg)
```
Comment (hkaiser) [inline libs/core/async_mpi/tests/unit/algorithm_transform_mpi.cpp:216]: We do have this: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/string_util/include/hpx/string_util/case_conv.hpp.

File: cmake/HPX_PerformCxxFeatureTests.cmake [hunk 16]
```diff
@@ -152,6 +152,8 @@ function(hpx_perform_cxx_feature_tests)
 
     hpx_check_for_cxx20_std_bit_cast(DEFINITIONS HPX_HAVE_CXX20_STD_BIT_CAST)
 
+    hpx_check_for_cxx20_std_identity(DEFINITIONS HPX_HAVE_CXX20_STD_BIT_CAST)
```
Comment (hkaiser) [inline cmake/HPX_PerformCxxFeatureTests.cmake:155]:
```suggestion
    hpx_check_for_cxx20_std_identity(DEFINITIONS 
HPX_HAVE_CXX20_STD_IDENTITY)
```

**Comment by isidorostsa** [comment]
@hkaiser Should I rebase on top of https://github.com/STEllAR-GROUP/hpx/pull/6507?

**Comment by isidorostsa** [comment]
Just rebased on master

**Comment by hkaiser** [comment]
Thanks so much!

**Comment by Pansysk75** [comment]
👏 👏
